File: backend/src/modules/auth/domain/ports/user_repository.py
from typing import Protocol

from src.modules.auth.domain.entities.user import User


class UserRepository(Protocol):
    """Port for user persistence operations."""

    def exists_by_email(self, email: str) -> bool:
        """Return whether a user already exists with the provided email."""

    def save(self, user: User) -> User:
        """Persist and return a user."""

    def find_by_email(self, email: str) -> User | None:
        """Return a user by email if it exists."""

    def find_by_id(self, user_id: str) -> User | None:
        """Return a user by identifier if it exists."""

    def update(self, user: User) -> User:
        """Update and return a persisted user."""

File: backend/src/modules/users/domain/ports/user.repository.ts
import { UserEntity } from '../../infra/entities/user.entity';

export interface IUserRepository {
  save(user: UserEntity): Promise<UserEntity>;
  findById(id: string): Promise<UserEntity | null>;
  findByEmail(email: string): Promise<UserEntity | null>;
  update(user: UserEntity): Promise<UserEntity>;
  delete(id: string): Promise<void>;
}

export const USER_REPOSITORY = 'USER_REPOSITORY';
